File: examples/custom_llm.py
"""Example: Creating a custom LLM backend.

Shows three ways to integrate any LLM with TreeDex:
1. FunctionLLM — wrap a callable
2. BaseLLM subclass — full control
3. LiteLLM — 100+ providers
"""

from treedex import TreeDex, BaseLLM, FunctionLLM


# --- Method 1: FunctionLLM ---
# Wrap any function that takes str and returns str.

def my_api_call(prompt: str) -> str:
    """Simulate an API call."""
    # Replace with your actual API call
    import urllib.request
    import json

    data = json.dumps({
        "model": "my-model",
        "messages": [{"role": "user", "content": prompt}],
    }).encode()

    req = urllib.request.Request(
        "https://my-api.example.com/v1/chat/completions",
        data=data,
        headers={"Content-Type": "application/json"},
    )

    with urllib.request.urlopen(req) as resp:
        body = json.loads(resp.read())

    return body["choices"][0]["message"]["content"]


# llm = FunctionLLM(my_api_call)


# --- Method 2: BaseLLM subclass ---
# Full control over initialization and generation.

class MyCustomLLM(BaseLLM):
    """Custom LLM that connects to your proprietary API."""

    def __init__(self, endpoint: str, token: str):
        self.endpoint = endpoint
        self.token = token

    def generate(self, prompt: str) -> str:
        import urllib.request
        import json

        data = json.dumps({"prompt": prompt}).encode()
        req = urllib.request.Request(
            self.endpoint,
            data=data,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self.token}",
            },
        )

        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read())["text"]


# llm = MyCustomLLM(endpoint="https://...", token="...")


# --- Method 3: Lambda (simplest) ---
# For quick prototyping.

# llm = FunctionLLM(lambda p: openai.chat(p))


# --- Usage ---
# index = TreeDex.from_file("doc.pdf", llm=llm)
# result = index.query("What is the main argument?")

print("See the code for three ways to create custom LLM backends!")
print("1. FunctionLLM(my_function)")
print("2. class MyLLM(BaseLLM): ...")
print("3. FunctionLLM(lambda p: ...)")

File: examples/node/custom-llm.ts
/**
 * Example: Creating a custom LLM backend (Node.js)
 *
 * Shows three ways to integrate any LLM with TreeDex:
 * 1. FunctionLLM — wrap a function
 * 2. BaseLLM subclass — full control
 * 3. Lambda — simplest
 */

import { BaseLLM, FunctionLLM } from "../../src/index.js";

// --- Method 1: FunctionLLM ---
// Wrap any function that takes string and returns string (or Promise<string>).

async function myApiCall(prompt: string): Promise<string> {
  const resp = await fetch("https://my-api.example.com/v1/chat/completions", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      model: "my-model",
      messages: [{ role: "user", content: prompt }],
    }),
  });

  const body = (await resp.json()) as {
    choices: Array<{ message: { content: string } }>;
  };
  return body.choices[0].message.content;
}

// const llm = new FunctionLLM(myApiCall);

// --- Method 2: BaseLLM subclass ---
// Full control over initialization and generation.

class MyCustomLLM extends BaseLLM {
  private endpoint: string;
  private token: string;

  constructor(endpoint: string, token: string) {
    super();
    this.endpoint = endpoint;
    this.token = token;
  }

  async generate(prompt: string): Promise<string> {
    const resp = await fetch(this.endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.token}`,
      },
      body: JSON.stringify({ prompt }),
    });

    const body = (await resp.json()) as { text: string };
    return body.text;
  }
}

// const llm = new MyCustomLLM("https://...", "token");

// --- Method 3: Lambda (simplest) ---

// const llm = new FunctionLLM((p) => someExistingApi(p));

// --- Usage ---
// const index = await TreeDex.fromFile("doc.pdf", llm);
// const result = await index.query("What is the main argument?");

console.log("See the code for three ways to create custom LLM backends!");
console.log("1. new FunctionLLM(myFunction)");
console.log("2. class MyLLM extends BaseLLM { ... }");
console.log("3. new FunctionLLM((p) => ...)");
